### ISSUE TITLE
dimulski - Admin can set the winner of the raffle to an address of his choosing

### ISSUE THREAD

**Issue by sherlock-admin3** [sherlock-audit/2024-08-winnables-raffles-judging#254]
dimulski

High

# Admin can set the winner of the raffle to an address of his choosing

### Summary

In the ``WinnablesPrizeManager.sol`` contract, an admin has to first lock some prize via the [lockNFT()](https://github.com/sherlock-audit/2024-08-winnables-raffles/blob/main/public-contracts/contracts/WinnablesPrizeManager.sol#L148-L165), [lockETH()](https://github.com/sherlock-audit/2024-08-winnables-raffles/blob/main/public-contracts/contracts/WinnablesPrizeManager.sol#L172-L188), or [lockTokens()](https://github.com/sherlock-audit/2024-08-winnables-raffles/blob/main/public-contracts/contracts/WinnablesPrizeManager.sol#L196-L213) functions, then a CCIP message will be sent to the ``WinnablesTicketManager.sol`` contract, and the admin can then fully set the raffle via the [createRaffle()](https://github.com/sherlock-audit/2024-08-winnables-raffles/blob/main/public-contracts/contracts/WinnablesTicketManager.sol#L252-L274) function. Once all the contracts are deployed the admin has to set the correct **contractAddress** and **chainSelector** via the [setCCIPCounterpart()](https://github.com/sherlock-audit/2024-08-winnables-raffles/blob/main/public-contracts/contracts/WinnablesPrizeManager.sol#L134-L140) function, so the ``WinnablesPrizeManager.sol`` contract [_ccipReceive()](https://github.com/sherlock-audit/2024-08-winnables-raffles/blob/main/public-contracts/contracts/WinnablesPrizeManager.sol#L260-L278) function can't be called by anyone, but only from an approved address on a specific chain. In the readme of the contest it is specifically said that the admin is not fully trusted: *Winnables admins cannot do anything to prevent a winner from withdrawing their prize* & *However it is not expected that the admin can steal funds that should have ended in a raffle participant’s wallet in any conceivable way*. However these core protocol invariants can be easily broken by a malicious admin. As can be seen from the code snippet below an admin can approve multiple contract on different chains to call the [_ccipReceive()](https://github.com/sherlock-audit/2024-08-winnables-raffles/blob/main/public-contracts/contracts/WinnablesPrizeManager.sol#L260-L278) function via the [setCCIPCounterpart()](https://github.com/sherlock-audit/2024-08-winnables-raffles/blob/main/public-contracts/contracts/WinnablesPrizeManager.sol#L134-L140) function:
```solidity
    function setCCIPCounterpart(
        address contractAddress,
        uint64 chainSelector,
        bool enabled
    ) external onlyRole(0) {
        _setCCIPCounterpart(contractAddress, chainSelector, enabled);
    }
```
A malicious admin can utilize this to set up a contract which sends a message to the ``WinnablesPrizeManager.sol`` contract [_ccipReceive()](https://github.com/sherlock-audit/2024-08-winnables-raffles/blob/main/public-contracts/contracts/WinnablesPrizeManager.sol#L260-L278) function and sets a winner for a specific raffle. Thus the admin can steal the promised rewards for the actual winner of the raffle, and the winner will get absolutely nothing in return, however the admin can still withdraw the accumulated tokens from the ticket sales from the ``WinnablesTicketManager.sol`` contract. The admin can also utilize a malicious contract to call the [WinnablesPrizeManager.sol::_ccipReceive()](https://github.com/sherlock-audit/2024-08-winnables-raffles/blob/main/public-contracts/contracts/WinnablesPrizeManager.sol#L260-L278) function, and cancel a raffle, thus freeing the funds he locked when he created the raffle. If for example he locked **1e18 ETH** as a prize for raffle 1, and approves a malicious contract to call [WinnablesPrizeManager.sol::_ccipReceive()](https://github.com/sherlock-audit/2024-08-winnables-raffles/blob/main/public-contracts/contracts/WinnablesPrizeManager.sol#L260-L278), then he uses the malicious contract to call the [WinnablesPrizeManager.sol::_ccipReceive()](https://github.com/sherlock-audit/2024-08-winnables-raffles/blob/main/public-contracts/contracts/WinnablesPrizeManager.sol#L260-L278) and cancels raffle 1, the admin can withdraw **1e18 ETH** from the contract. When a winner is drawn and set and tries to call the [claimPrize()](https://github.com/sherlock-audit/2024-08-winnables-raffles/blob/main/public-contracts/contracts/WinnablesPrizeManager.sol#L105-L124) function and claim his prize, he won't be able to as there won't be any ETH in the contract. 
### Root Cause

A malicious admin can approve multiple contracts to call the [_ccipReceive()](https://github.com/sherlock-audit/2024-08-winnables-raffles/blob/main/public-contracts/contracts/WinnablesPrizeManager.sol#L260-L278) function via the [setCCIPCounterpart()](https://github.com/sherlock-audit/2024-08-winnables-raffles/blob/main/public-contracts/contracts/WinnablesPrizeManager.sol#L134-L140) function .

### Internal pre-conditions

1. Admin creates a raffle
2. Users start buying tickets via the [buyTickets()](https://github.com/sherlock-audit/2024-08-winnables-raffles/blob/main/public-contracts/contracts/WinnablesTicketManager.sol#L182-L210) function, but a winner is not yet drawn

### Attack Path

1. Admin creates a raffle with ID 1
2. Admin approves a contract he controls to successfully send messages to the ``WinnablesPrizeManager.sol`` contract [_ccipReceive()](https://github.com/sherlock-audit/2024-08-winnables-raffles/blob/main/public-contracts/contracts/WinnablesPrizeManager.sol#L260-L278) function
3. As users are buying tickets for raffle with ID 1, the malicious admin can utilize the contract he approved in step 2 to send a CCIP message to ``WinnablesPrizeManager.sol`` contract [_ccipReceive()](https://github.com/sherlock-audit/2024-08-winnables-raffles/blob/main/public-contracts/contracts/WinnablesPrizeManager.sol#L260-L278) function, which sets the address of the winner for the raffle with ID 1 to an address that the attacker controls. lets call it **address A**. 
4. Now  **address A**  can call the [claimPrize()](https://github.com/sherlock-audit/2024-08-winnables-raffles/blob/main/public-contracts/contracts/WinnablesPrizeManager.sol#L105-L124) function and claim the prize that the admin initially locked. 
```solidity
    function claimPrize(uint256 raffleId) external {
        ...
        if (msg.sender != rafflePrize.winner) revert UnauthorizedToClaim();
        if (rafflePrize.status == RafflePrizeStatus.CLAIMED) revert AlreadyClaimed();
        rafflePrize.status = RafflePrizeStatus.CLAIMED;
        ...
    }
```  
As can be seen from the code snippet below the [claimPrize()](https://github.com/sherlock-audit/2024-08-winnables-raffles/blob/main/public-contracts/contracts/WinnablesPrizeManager.sol#L105-L124) function sets the **rafflePrize.status** to **CLAIMED**, so this function will always revert when called again. 
5. When there have been enough users, and a winner can be drawn for the raffle with ID 1, the attacker or the unsuspecting users can call the [propagateRaffleWinner()](https://github.com/sherlock-audit/2024-08-winnables-raffles/blob/main/public-contracts/contracts/WinnablesTicketManager.sol#L334-L345) function, so the AVAX that was accumulated from the tickets that the users bought can be freely withdrawn by the admin via the [withdrawETH()](https://github.com/sherlock-audit/2024-08-winnables-raffles/blob/main/public-contracts/contracts/WinnablesTicketManager.sol#L300-L306). As mentioned in step 4, when the newly set winner for the raffle with ID 1, tries to call the [claimPrize()](https://github.com/sherlock-audit/2024-08-winnables-raffles/blob/main/public-contracts/contracts/WinnablesPrizeManager.sol#L105-L124) function in order to claim his prize, the function will revert. 

### Impact

A malicious admin can set the winner of a certain raffle, and in this way deny the winner of the raffle from withdrawing his prize. Or cancel the raffle only in ``WinnablesPrizeManager.sol``, and withdraw the funds he locked when he initially created the raffle. At the same time he can also withdraw the proceeds from the sold tickets in the ``WinnablesTicketManager.sol`` contract. Thus breaking a couple of protocol invariants and effectively stealing money from the users, thus the high severity. 

### Mitigation

Allow the admin to call the [setCCIPCounterpart()](https://github.com/sherlock-audit/2024-08-winnables-raffles/blob/main/public-contracts/contracts/WinnablesPrizeManager.sol#L134-L140) function only once, and set the correct parameters.

Duplicate of #277

**Comment by AtanasDimulski** [comment]
Escalate,
Since there are no comments provided as to why the judge believes this issue is invalid, I am guessing it is because the judge has assumed that the admin is trusted, however in the README it is specified that admins are not fully trusted: *Winnables admins cannot do anything to prevent a winner from withdrawing their prize & However it is not expected that the admin can steal funds that should have ended in a raffle participant’s wallet in any conceivable way*.

**Comment by sherlock-admin3** [comment]
> Escalate,
> Since there are no comments provided as to why the judge believes this issue is invalid, I am guessing it is because the judge has assumed that the admin is trusted, however in the README it is specified that admins are not fully trusted: *Winnables admins cannot do anything to prevent a winner from withdrawing their prize & However it is not expected that the admin can steal funds that should have ended in a raffle participant’s wallet in any conceivable way*.

You've created a valid escalation!

To remove the escalation from consideration: Delete your comment.

You may delete or edit your escalation comment anytime before the 48-hour escalation window closes. After that, the escalation becomes final.

**Comment by mystery0x** [comment]
All admin related issues have been discussed with the HOJ and the sponsor who wished the following governing statements could have been more carefully written and/or omitted:

"The protocol working as expected relies on having an admin creating raffles. It should be expected that the admin will do their job. However it is not expected that the admin can steal funds that should have ended in a raffle participant’s wallet in any conceivable way."

Much as many of the findings suggest good fixes, the conclusion is that they will all be deemed low unless it's beyond the trusted admin control.

**Comment by philbugcatcher** [comment]
@mystery0x The README language leaves no ambiguity when it states:
> it is not expected that the admin can steal funds that shold have ended in a raffle participant's wallet in any conceivable way.

The phrase “any conceivable way” is crucial here.

As demonstrated in @AtanasDimulski issue and the issues that I've raised (https://github.com/sherlock-audit/2024-08-winnables-raffles-judging/issues/63 and https://github.com/sherlock-audit/2024-08-winnables-raffles-judging/issues/62), there are very conceivable ways this claim can be violated.

The admin has even acknowledged in the chat that issues of this nature should be considered valid.

There are several viable solutions that would preserve the admin’s control over protocol maintenance while preventing this vulnerability. For instance, implementing a mapping from the CCIP counterpart used for a specific raffle, and restricting prize unlocking to that counterpart, would be a straightforward fix.

Ultimately, while it is up to the admin to decide whether or not to address this issue, it remains our responsibility as auditors to clearly document the existence of the vulnerability in our report.

**Comment by AtanasDimulski** [comment]
Hey @mystery0x, thanks for your feedback. However I am not aware of what you have discussed with the sponsor and HOJ, and frankly, it doesn't matter. Sherlock rules are pretty clear: *If the protocol team provides specific information in the README or CODE COMMENTS, that information stands above all judging rules. In case of contradictions between the README and CODE COMMENTS, the README is the chosen source of truth.* & *Direct Protocol Owner/Admin rug pulls. If a protocol specifically mentions the restrictions imposed on the owner/admin, issues describing an attack that results in bypassing these restrictions can be considered valid.*. If the sponsor wanted to clarify the government statements he had time to do so, as Sherlock allows the README to be updated up to 24 hours before the contest ends. In the README it is clearly stated that:

**The principles that must always remain true are:**
 - Winnables admins cannot do anything to prevent a winner from withdrawing their prize
 - Participants in a raffle that got cancelled can always get refunded
 - Admins cannot affect the odds of a raffle
 - It is not expected that the admin can steal funds that should have ended in a raffle participant’s wallet in any conceivable way.

Not accepting those issues goes against Sherlock rules. Thanks for your time

**Comment by WangSecurity** [comment]
After additionally considering this issue, the escalation will be accepted and the issue will be validated with medium severity and duplicated with #277. For more details, look into the discussion under #277.

**Comment by WangSecurity** [comment]
Result:
Medium
Duplicate of #277

**Comment by sherlock-admin4** [comment]
Escalations have been resolved successfully!

Escalation status:
- [AtanasDimulski](https://github.com/sherlock-audit/2024-08-winnables-raffles-judging/issues/254/#issuecomment-2329486292): accepted
